rust/src/coordinator.rs

```diff
@@ -162,89 +120,107 @@ impl<N> Default for Coordinator<N> {
             sum_dict,
             seed_dict,
             mask_dict,
-            model,
-            masked_model,
             events,
         }
     }
 }
 
-pub trait Coordinators: Sized {
-    define_trait_fields!(
-        pk, CoordinatorPublicKey;
-        sk, CoordinatorSecretKey;
-        sum, f64;
-        update, f64;
-        min_sum, usize;
-        min_update, usize;
-        seed, RoundSeed;
-        phase, Phase;
-        sum_dict, SumDict;
-        seed_dict, SeedDict;
-        mask_dict, MaskDict;
-        masked_model, Option<MaskedModel>;
-        events, VecDeque<ProtocolEvent>;
-    );
-
+impl Coordinator {
     /// Create a coordinator. Fails if there is insufficient system entropy to generate secrets.
-    fn new() -> Result<Self, InitError>;
+    pub fn new() -> Result<Self, InitError> {
+        // crucial: init must be called before anything else in this module
+        sodiumoxide::init().or(Err(InitError))?;
+        let seed = randombytes(32);
+        Ok(Self {
+            seed,
+            ..Default::default()
+        })
+    }
+
+    /// Emit an event
+    pub fn emit_event(&mut self, event: ProtocolEvent) {
+        self.events.push_back(event);
+    }
 
-    /// Emit an event.
-    fn emit_event(&mut self, event: ProtocolEvent) {
-        self.events_mut().push_back(event);
+    /// Retrieve the next event
+    pub fn next_event(&mut self) -> Option<ProtocolEvent> {
+        self.events.pop_front()
     }
 
-    /// Retrieve the next event.
-    fn next_event(&mut self) -> Option<ProtocolEvent> {
-        self.events_mut().pop_front()
+    fn message_open(&self) -> MessageOpen<'_, '_> {
+        MessageOpen {
+            recipient_pk: &self.pk,
+            recipient_sk: &self.sk,
+        }
     }
 
     /// Validate and handle a sum, update or sum2 message.
-    fn handle_message(&mut self, bytes: &[u8]) -> Result<(), PetError> {
-        match self.phase() {
-            Phase::Idle => Err(PetError::InvalidMessage),
-            Phase::Sum => self.handle_sum_message(bytes),
-            Phase::Update => self.handle_update_message(bytes),
-            Phase::Sum2 => self.handle_sum2_message(bytes),
+    pub fn handle_message(&mut self, bytes: &[u8]) -> Result<(), PetError> {
+        let message = self
+            .message_open()
+            .open(&bytes)
```

Comment:
this is somewhat confusing to read, looks like it is opened twice. but actually the first occurrence refers to the actor and the second to the action. basically coming back to the naming discussion.

Reply:
I agree, it weirded me out too when I wrote it. I've open a ticket for the renaming because I wouldn't want to hold the PR further on this, though.